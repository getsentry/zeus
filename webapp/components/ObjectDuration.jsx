import React, {Component} from 'react';
import PropTypes from 'prop-types';
import percentile from 'percentile';

import Duration from './Duration';

const ONE_SECOND_IN_MS = 1000;

export class AggregateDuration extends Component {
  static propTypes = {
    data: PropTypes.arrayOf(
      PropTypes.shape({
        started_at: PropTypes.string,
        finished_at: PropTypes.string,
        duration: PropTypes.number
      })
    ).isRequired
  };

  constructor(props, context) {
    super(props, context);
    this.state = {
      duration: this.getDuration(props.data)
    };
  }

  componentDidMount() {
    this.setDurationTicker();
  }

  componentWillUnmount() {
    if (this.ticker) {
      clearTimeout(this.ticker);
      this.ticker = null;
    }
  }

  setDurationTicker() {
    let {data} = this.props;
    if (data.find(d => !d.finished_at)) {
      this.ticker = setTimeout(() => {
        this.setState({
          duration: this.getDuration(data)
        });
        this.setDurationTicker();
      }, ONE_SECOND_IN_MS);
    }
  }

  getDuration(data) {
    if (!data) return null;
    let values = data
      .map(item => {
        if (!item) return null;
        if (item.duration) return item.duration;
        if (!item.finished_at) {
          if (item.started_at) {
            return new Date().getTime() - new Date(item.started_at).getTime();
          }
          return null;
        }
        return new Date(item.finished_at).getTime() - new Date(item.started_at).getTime();
      })
      .filter(i => !!i);
    if (!values.length) return null;
    return percentile(95, values);
  }

  render() {
    let {duration} = this.state;
    if (duration === null) return null;
    return <Duration ms={duration} {...this.props} />;
  }
}

export default class ObjectDuration extends Component {
  static propTypes = {
    data: PropTypes.shape({
      started_at: PropTypes.string,
      finished_at: PropTypes.string,
      duration: PropTypes.number
    }).isRequired
  };

  render() {
    return <AggregateDuration data={[this.props.data]} />;
  }
}
