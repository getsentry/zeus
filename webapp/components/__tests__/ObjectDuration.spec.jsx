import React from 'react';
import {render} from 'enzyme';

import ObjectDuration from '../ObjectDuration';

describe('ObjectDuration', () => {
  it('renders unfinished', () => {
    const tree = render(
      <ObjectDuration
        data={{
          started_at: null,
          finished_at: null
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders in-progress', () => {
    const tree = render(
      <ObjectDuration
        data={{
          started_at: '2017-10-17T04:41:26',
          finished_at: null
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders finished', () => {
    const tree = render(
      <ObjectDuration
        data={{
          started_at: '2017-10-17T04:41:20',
          finished_at: '2017-10-17T04:43:31'
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
