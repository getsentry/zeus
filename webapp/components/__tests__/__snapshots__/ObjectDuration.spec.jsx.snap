// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectDuration renders finished 1`] = `
<span>
  2m
</span>
`;

exports[`ObjectDuration renders in-progress 1`] = `
<span>
  6s
</span>
`;

exports[`ObjectDuration renders unfinished 1`] = `null`;
